### PR TITLE
setup xtask infra for cheatsheets

### PR DESCRIPTION
**Pull request by miguelraz**
## Problem:

 Making cheatsheets like "From Python to Rust" and friends is going to be a maintainability nightmare as `training-materials` grows, changes ownership, etc., over time. This was raised as a concern in #148 and this PR seeks to fix that.

## Proposed solution:

Setup a `xtask` style workflow where we 

* call `cargo xtask make-cheatsheet python` at the root folder
* scrape Markdown headers in `SUMMARY..md` and segment topics by `Rust Fundamentals`, `Applied Rust` and `Advanced Rust`
* write out to `src/python-cheatsheet.md` if it doesn't exist
* if it does exist, check that it in sync: all headers in `python-cheatsheet.md` are in the appropriate sections, in order, and none are missing.

The code is (I think) defensive to spare future pain for someone else looking at this code ~~(myself in a few months)~~.

This PR only sets up the infrastructure but doesn't add any cheatsheets themselves - I'll file a separate PR with a basic Julia cheatsheet once this lands to separate concerns.

-------

<details>
<summary> Current functionality: </summary>

```text
❯ cargo xtask make-cheatsheet python
   Compiling xtask v0.1.0 (C:\Users\mrg\work\rust-training\xtask)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 1.10s
     Running `target\debug\xtask.exe make-cheatsheet python`
Cheatsheat for python written at ./training-slides/src/python-cheatsheet.md

❯ cat .\training-slides\src\python-cheatsheet.md
# Rust Fundamentals
## Overview
## Installation
## Basic Types
## Control Flow
## Compound Types
## Ownership and Borrowing
## Error Handling
## Collections
## Iterators
## Imports and Modules
## Good Design Practices

# Applied Rust
## Methods and Traits
## Rust I/O Traits
## Generics
## Lifetimes
## Cargo Workspaces
## Heap Allocation (Box and Rc)
## Shared Mutability (Cell, RefCell)
## Thread Safety (Send/Sync, Arc, Mutex)
## Closures and the Fn/FnOnce/FnMut traits
## Spawning Threads and Scoped Threads

# Advanced Rust
## Advanced Strings
## Debugging Rust
## Dependency Management with Cargo
## Deref Coercions
## Design Patterns
## Documentation
## Drop, Panic and Abort
## Dynamic Dispatch
## Macros
## Property Testing
## Rust Projects Build Time
## Send and Sync
## Serde
## Testing
## The stdlib
## Using Cargo
## Using Types to encode State
## WASM

❯ cargo xtask make-cheatsheet python
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.04s
     Running `target\debug\xtask.exe make-cheatsheet python`
File python-cheatsheet.md already exists - checking it's in sync
Neat! python-cheatsheet.md is in sync

rust-training on  xtask-cheatsheet [$?] via 🦀 v1.78.0
❯ cargo xtask test-cheatsheet python
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.04s
     Running `target\debug\xtask.exe test-cheatsheet python`
Neat! python-cheatsheet.md is in sync
```

</details>

### CODE REVIEW

**Comment by jonathanpallant** [comment]
Could we add a few words (here, or in a README) about how the cheatsheets work - to save having to read and understand the source code?

**Comment by jonathanpallant** [comment]
Also, it needs a rebase on `main` to fix the CloudFlare CI failure, and some thought about the workspaces to fix the GHA CI failure.

**Comment by miguelraz** [comment]
@jonathanpallant 

* added a README.md
* rebased to main to have Cloudflare deploy passing
* fixed workspace issue by adding a `exclude = ["examples/"]` array to the `Cargo.toml` and CI is happy

File: xtask/README.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+...
+## Generics
+...
+```
+
+per programming language that match each slide we have in our normal syllabus with an entry on the second header (`## Methods and Traits`, for example) level.
+
+## How
+
+As `training-material` grows and changes, maintenance could be a nightmare. We basically don't want to be in the business of remember that a certain slide got reordered or moved from one section to another and thus needs changing in all the cheathseets as well. Therefore, this tool seeks to alleviate that with the following workflow:
```
Comment (jonathanpallant) [inline xtask/README.md:26]:
```suggestion
As `training-material` grows and changes, maintenance could be a nightmare. We basically don't want to be in the business of remembering that a certain slide got reordered or moved from one section to another and thus needs changing in all the cheatsheets as well. Therefore, this tool seeks to alleviate that with the following workflow:
```

File: xtask/README.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+* call `cargo xtask make-cheatsheet python` at the root folder
+* scrape Markdown headers in `SUMMARY.md` and segment topics by `Rust Fundamentals`, `Applied Rust` and `Advanced Rust`
```
Comment (jonathanpallant) [inline xtask/README.md:29]: Who does this - the user or the tool? Because step 1 in this list is something the user has to do.

File: xtask/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+use std::env;
+
+// Code adapted from the xtask workflow in rust-exercises
+fn main() -> color_eyre::Result<()> {
+    color_eyre::install()?;
+
+    // first arg is the name of the executable; skip it
+    let args = env::args().skip(1).collect::<Vec<_>>();
+    let args = args.iter().map(|arg| &arg[..]).collect::<Vec<_>>();
```
Comment (jonathanpallant) [inline xtask/src/main.rs:13]: What is this line for? `args` is already a `Vec<String>` right?

File: xtask/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    let langs = vec!["python", "go", "cpp", "swift", "java", "julia", "c"];
+
+    if !langs.contains(&args[1]) {
+        panic!("{} is not a valid language name. \nExpected one of: python, go, cpp, swift, java, julia, c", &args[1]);
```
Comment (jonathanpallant) [inline xtask/src/main.rs:18]: I'd print the lang list in the panic so it doesn't get out of sync.

File: xtask/src/tasks.rs [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+        .rfind(']')
+        .expect("the markdown file entry did not have a ']'");
+    let bot = line
+        .find('[')
+        .expect("the markdown file entry did not have a '['");
+    String::from(&line[bot + 1..top])
+}
+
+#[test]
+fn test_get_slide_name() {
```
Comment (jonathanpallant) [inline xtask/src/tasks.rs:28]: Tests should go in a `#[cfg(test)] mod test { ... }`

File: xtask/src/tasks.rs [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    let mut current_section: Vec<String> = Vec::new();
+
+    if !text.contains(INITIAL_HEADER) {
+        panic!("Your INITIAL_HEADER is not part of the input. Check your `SUMMARY.md` for {INITIAL_HEADER}");
+    }
+    if !text.contains(LAST_HEADER) {
+        panic!("YOUR LAST_HEADER is not part of the text input. CHECK your `SUMMARY.md` for {LAST_HEADER}");
+    }
+
+    let first_header = text.find(INITIAL_HEADER).unwrap();
```
Comment (jonathanpallant) [inline xtask/src/tasks.rs:52]: You can handle the error here rather than having a separate `contains` check which might get isolated from this line.

```rust
let Some(first_header) = text.find(INITIAL_HEADER) else {
    panic!("Could not find the initial header {INITIAL_HEADER:?}. Check your `SUMMARY.md`");
}
```

File: xtask/src/tasks.rs [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+    if !current_section.is_empty() {
+        result.push(current_section);
+    }
+
+    result
+}
+
+#[test]
+fn test_focus_regions() {
```
Comment (jonathanpallant) [inline xtask/src/tasks.rs:80]: Tests should go in a #[cfg(test)] mod test { ... }

File: xtask/src/tasks.rs [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+        .collect::<Vec<String>>();
+
+    SlidesSection {
+        header,
+        slide_titles,
+    }
+}
+
+#[test]
+fn test_extract_slides() {
```
Comment (jonathanpallant) [inline xtask/src/tasks.rs:140]: Tests should go in a #[cfg(test)] mod test { ... }

**Comment by jonathanpallant** [comment]
OK, I think this tool makes sense. I had some minor points about the program itself, but I also think we should be running the unit tests in CI, and testing our cheat-sheets in CI.

File: xtask/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+// Code adapted from the xtask workflow in rust-exercises
+fn main() -> color_eyre::Result<()> {
+    color_eyre::install()?;
+
+    // first arg is the name of the executable; skip it
+    let args = env::args().skip(1).collect::<Vec<_>>();
+    let args = args.iter().map(|arg| &arg[..]).collect::<Vec<_>>();
+
+    let langs = vec!["python", "go", "cpp", "swift", "java", "julia", "c"];
```
[the remark this answers is not in the document]
Comment (miguelraz) [reply xtask/src/main.rs:15]: Yup!

**Comment by jonathanpallant** [comment]
re-assign to me when you're ready, thanks!

File: xtask/README.md (hunk 2 above)
Comment (miguelraz) [reply xtask/README.md:29]: Yes, for now it's the user's responsibility to run the check after modifying any of the cheatsheets - I wouldn't mind doing a follow up PR that adds it to the build scripts. I'll add that as a follow up task for myself on Clickup.

File: xtask/src/main.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (miguelraz) [reply xtask/src/main.rs:15]: Done.

File: xtask/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    if !langs.contains(&args[1]) {
+        panic!("{} is not a valid language name. \nExpected one of: python, go, cpp, swift, java, julia, c", &args[1]);
+    }
+
+    match &args[..] {
+        ["make-cheatsheet", lang] => tasks::make_cheatsheet(lang),
+        ["test-cheatsheet", lang] => tasks::test_cheatsheet(lang),
+        _ => {
+            eprintln!(
+                "cargo xtask
```
[the remark this answers is not in the document]
Comment (miguelraz) [reply xtask/src/main.rs:26]: Good call, done.

File: xtask/src/main.rs (hunk 4 above)
Comment (miguelraz) [reply xtask/src/main.rs:18]: Good call, added that and some cleaner formatting in the print.

File: xtask/src/tasks.rs (hunk 5 above)
Comment (miguelraz) [reply xtask/src/tasks.rs:28]: Neater, done.

File: xtask/src/tasks.rs (hunk 7 above)
Comment (miguelraz) [reply xtask/src/tasks.rs:80]: Also refactored, thanks.

File: xtask/src/tasks.rs (hunk 8 above)
Comment (miguelraz) [reply xtask/src/tasks.rs:140]: Also refactored.

File: xtask/src/tasks.rs (hunk 6 above)
Comment (miguelraz) [reply xtask/src/tasks.rs:52]: Oh, that is a much nicer form. Done.

**Comment by miguelraz** [comment]
Force pushed to get rid of a pesky rebase muddying the history horribly.

**Comment by miguelraz** [comment]
Should be ready for review again.

File: xtask/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+USAGE:
+    cargo xtask [COMMAND]
+
+COMMANDS:
+    make-cheatsheet [LANG]      make LANG cheatsheet by scraping slides names in `SUMMARY.md`
+    test-cheatsheet [LANG]      test LANG's cheatsheet (all `SUMMARY.md` items are in sheet and viceversa)
+
+LANG:
+    Valid values are `python, go, cpp, swift, java, julia, c`
```
Comment (jonathanpallant) [inline xtask/src/main.rs:17]: I would replace this list with a magic token like `$$LANG_LIST$$` and do `HELP_TEXT.replace("$$LANG_LIST$$", langs.join(", ")` this string before you print it, like:

```rust
static HELP_TEXT: &'static str = "cargo xtask

LANG:

  We support $$LANG_LIST$$
";

fn main() {
    let lang_list = ["Apples", "Bananas", "Pears"];
    println!("{}", HELP_TEXT.replace("$$LANG_LIST$$", &join_str(&lang_list)));
}

fn join_str(input: &[&str]) -> String {
    let mut output = String::new();
    let mut iter = input.iter().peekable();
    while let Some(lang) = iter.next() {
        if iter.peek().is_none() {
            output += " and ";
            output += lang;
        } else {
            output += ", ";
            output += lang;
        }
    }
    output
}
```

**Comment by jonathanpallant** [comment]
There's quite a lot of code here, and coming to it fresh after vacation I lacked any real context. Some file-level comments explaining the purpose of the file and some function-level comments explaining the purpose of each function would be a useful addition.

File: xtask/src/main.rs (hunk 11 above)
Comment (jonathanpallant) [reply xtask/src/main.rs:17]: the fancy join is optional

**Comment by miguelraz** [comment]
I'll write those in later today, good point.

File: xtask/src/main.rs (hunk 11 above)
Comment (miguelraz) [reply xtask/src/main.rs:17]: Fancy indeed. I'll work that in.

**Comment by miguelraz** [comment]
@jonathanpallant this feels ready for review again.

**Review by jonathanpallant — CHANGES_REQUESTED** [review]

**Comment by jonathanpallant** [comment]
I elected to push some changes rather than leave a bunch of review comments. Please do review the diff and let me know if you have any questions.

The final matter is deciding how we want CI to run this, to automatically check out cheatsheets. Ideally we'd have a `cargo xtask test-cheatsheets` which automatically tested all of them.

**Comment by miguelraz** [comment]
All the comments being hoisted to a README make sense to me and the added level of doc comments is welcome, thanks!

I'm in favor of writing a `cargo xtask test-all`.

**Comment by miguelraz** [comment]
Resolved a small merge conflict by accepting the `xtask` `target` folder as part of `.gitignore`

**Comment by miguelraz** [comment]
Ah, I noticed there's a corner case I haven't handled, I'm going to work on it for a bit.
